File: go/tinyrwm.go
package main

import (
	"context"
	"errors"
	"fmt"
	"log"
	"os"
	"os/exec"
	"slices"

	"codeberg.org/river/tinyrwm/go/internal/proto"
	"hazelnut.eclair.cafe/wlcl"
)

type Output struct {
	proto.RiverOutputV1Stub

	ID proto.RiverOutputV1

	Removed bool
}

func (o *Output) HandleRiverOutputV1Removed(ctx context.Context) {
	o.Removed = true
}

func (o *Output) MaybeDestroy() bool {
	if !o.Removed {
		return false
	}

	o.ID.Destroy()
	return true
}

func NewOutput(id proto.RiverOutputV1) *Output {
	output := &Output{
		ID: id,
	}

	id.SetUserData(output)
	return output
}

type Window struct {
	proto.RiverWindowV1Stub

	ID   proto.RiverWindowV1
	Node proto.RiverNodeV1

	X, Y          int32
	Width, Height int32

	New    bool
	Closed bool

	PointerMoveRequested *Seat

	PointerResizeRequested      *Seat
	PointerResizeRequestedEdges uint32
}

func (w *Window) SetPosition(x, y int32) {
	w.Node.SetPosition(x, y)
	w.X, w.Y = x, y
}

func (w *Window) HandleRiverWindowV1Closed(ctx context.Context) {
	w.Closed = true
}

func (w *Window) HandleRiverWindowV1Dimensions(ctx context.Context, width int32, height int32) {
	w.Width, w.Height = width, height
}

func (w *Window) HandleRiverWindowV1PointerMoveRequested(ctx context.Context, seat proto.RiverSeatV1) {
	w.PointerMoveRequested = seat.UserData().(*Seat)
}

func (w *Window) HandleRiverWindowV1PointerResizeRequested(ctx context.Context, seat proto.RiverSeatV1, edges uint32) {
	w.PointerResizeRequested = seat.UserData().(*Seat)
	w.PointerResizeRequestedEdges = edges
}

func (w *Window) MaybeDestroy() bool {
	if !w.Closed {
		return false
	}

	w.Node.Destroy()
	w.ID.Destroy()
	return true
}

func (w *Window) Manage() {
	if w.New {
		w.New = false
		w.SetPosition(0, 0)
		w.ID.ProposeDimensions(0, 0)
	}

	if w.PointerMoveRequested != nil {
		w.PointerMoveRequested.PointerMove(w)
		w.PointerMoveRequested = nil
	}

	if w.PointerResizeRequested != nil {
		w.PointerResizeRequested.PointerResize(w, w.PointerResizeRequestedEdges)
		w.PointerResizeRequested = nil
	}
}

func NewWindow(id proto.RiverWindowV1) *Window {
	window := &Window{
		ID:   id,
		Node: id.GetNode(),
		New:  true,
	}

	id.SetUserData(window)
	return window
}

type SeatOp interface {
	Manage(state *SeatOpState)
	Render(state *SeatOpState)
	InformStart(w *Window)
	InformEnd(w *Window)
}

type SeatOpState struct {
	Window *Window
	Dx, Dy int32
}

type Seat struct {
	proto.RiverSeatV1Stub

	WM *WindowManager

	ID              proto.RiverSeatV1
	XkbBindings     []*XkbBinding
	PointerBindings []*PointerBinding

	Focused    *Window
	Hovered    *Window
	Interacted *Window

	PendingAction func()

	Op         SeatOp
	OpState    SeatOpState
	OpReleased bool

	New     bool
	Removed bool
}

func (s *Seat) Focus(w *Window) {
	wm := s.WM

	if w == nil && len(wm.Windows) > 0 {
		w = wm.Windows[len(wm.Windows)-1]
	}

	if s.Focused == w {
		return
	}

	if w != nil {
		s.ID.FocusWindow(w.ID)

		// Raise to top
		idx := slices.Index(wm.Windows, w)
		wm.Windows = append(wm.Windows[:idx], wm.Windows[idx+1:]...)
		wm.Windows = append(wm.Windows, w)

		w.Node.PlaceTop()
	} else {
		s.ID.ClearFocus()
	}

	s.Focused = w
}

func (s *Seat) HandleRiverSeatV1OpDelta(ctx context.Context, dx int32, dy int32) {
	s.OpState.Dx, s.OpState.Dy = dx, dy
}

func (s *Seat) HandleRiverSeatV1OpRelease(ctx context.Context) {
	s.OpReleased = true
}

func (s *Seat) HandleRiverSeatV1PointerEnter(ctx context.Context, window proto.RiverWindowV1) {
	s.Hovered = window.UserData().(*Window)
}

func (s *Seat) HandleRiverSeatV1PointerLeave(ctx context.Context) {
	s.Hovered = nil
}

func (s *Seat) HandleRiverSeatV1Removed(ctx context.Context) {
	s.Removed = true
}

func (s *Seat) HandleRiverSeatV1WindowInteraction(ctx context.Context, window proto.RiverWindowV1) {
	s.Interacted = window.UserData().(*Window)
}

func (s *Seat) MaybeDestroy() bool {
	if !s.Removed {
		return false
	}

	for _, b := range s.XkbBindings {
		b.ID.Destroy()
	}
	for _, b := range s.PointerBindings {
		b.ID.Destroy()
	}

	s.ID.Destroy()
	return true
}

func (s *Seat) Manage() {
	if s.New {
		s.New = false
		for _, b := range s.XkbBindings {
			b.ID.Enable()
		}
		for _, b := range s.PointerBindings {
			b.ID.Enable()
		}
	}

	if w := s.Focused; w != nil && w.Closed {
		s.Focused = nil
	}

	// If no window was interacted with in the current manage sequence,
	// intentionally pass nil to ensure the window on top has focus.
	// This is necessary to handle new windows for example.
	s.Focus(s.Interacted)
	s.Interacted = nil

	if s.PendingAction != nil {
		s.PendingAction()
		s.PendingAction = nil
	}

	if op := s.Op; op != nil {
		switch {
		case s.OpReleased:
			op.InformEnd(s.OpState.Window)
			fallthrough
		case s.OpState.Window.Closed:
			s.ID.OpEnd()
			s.Op = nil
		default:
			op.Manage(&s.OpState)
		}
	}
	s.OpReleased = false
}

func (s *Seat) Render() {
	if op := s.Op; op != nil {
		op.Render(&s.OpState)
	}
}

func (s *Seat) StartOp(w *Window, op SeatOp) {
	if s.Op != nil {
		return
	}

	s.Op = op
	s.OpState = SeatOpState{
		Window: w,
	}

	s.Focus(w)

	s.ID.OpStartPointer()
	op.InformStart(w)
}

func (s *Seat) PointerMove(w *Window) {
	s.StartOp(w, NewSeatOpMove(w))
}

func (s *Seat) PointerResize(w *Window, edges uint32) {
	s.StartOp(w, NewSeatOpResize(w, edges))
}

func NewSeat(id proto.RiverSeatV1, wm *WindowManager) *Seat {
	const (
		// See xkbcommon-keysyms.h
		spaceSym = 0x0020
		nSym     = 0x006e
		qSym     = 0x0071
		escSym   = 0xff1b

		leftButton  = 0x110
		rightButton = 0x111
	)

	const super = proto.RiverSeatV1ModifiersMod4

	seat := &Seat{
		WM:  wm,
		ID:  id,
		New: true,
	}

	seat.XkbBindings = []*XkbBinding{
		NewXkbBinding(seat, spaceSym, super, func() {
			cmd := exec.Command("foot")
			if err := cmd.Start(); err != nil {
				log.Printf("Failed to launch foot: %v", err)
			} else {
				cmd.Process.Release()
			}
		}),
		NewXkbBinding(seat, nSym, super, func() {
			if len(seat.WM.Windows) > 0 {
				seat.Focus(seat.WM.Windows[0])
			}
		}),
		NewXkbBinding(seat, qSym, super, func() {
			if seat.Focused != nil {
				seat.Focused.ID.Close()
			}
		}),
		NewXkbBinding(seat, escSym, super, func() {
			seat.WM.WindowManagerV1.ExitSession()
		}),
	}

	seat.PointerBindings = []*PointerBinding{
		NewPointerBinding(seat, leftButton, super, func() {
			if seat.Hovered != nil {
				seat.PointerMove(seat.Hovered)
			}
		}),
		NewPointerBinding(seat, rightButton, super, func() {
			if seat.Hovered != nil {
				seat.PointerResize(seat.Hovered, proto.RiverWindowV1EdgesRight|proto.RiverWindowV1EdgesBottom)
			}
		}),
	}

	seat.ID.SetUserData(seat)
	return seat
}

type SeatOpMove struct {
	StartX, StartY int32
}

func (s *SeatOpMove) Manage(state *SeatOpState) {}

func (s *SeatOpMove) Render(state *SeatOpState) {
	state.Window.SetPosition(s.StartX+state.Dx, s.StartY+state.Dy)
}

func (s *SeatOpMove) InformStart(w *Window) {}

func (s *SeatOpMove) InformEnd(w *Window) {}

func NewSeatOpMove(w *Window) *SeatOpMove {
	return &SeatOpMove{
		StartX: w.X,
		StartY: w.Y,
	}
}

type SeatOpResize struct {
	StartX, StartY          int32
	StartWidth, StartHeight int32
	Edges                   uint32
}

func (s *SeatOpResize) Manage(state *SeatOpState) {
	w, h := s.StartWidth, s.StartHeight
	switch {
	case s.Edges&proto.RiverWindowV1EdgesLeft != 0:
		w -= state.Dx
	case s.Edges&proto.RiverWindowV1EdgesRight != 0:
		w += state.Dx
	}

	switch {
	case s.Edges&proto.RiverWindowV1EdgesTop != 0:
		h -= state.Dy
	case s.Edges&proto.RiverWindowV1EdgesBottom != 0:
		h += state.Dy
	}

	state.Window.ID.ProposeDimensions(max(w, 1), max(h, 1))
}

func (s *SeatOpResize) Render(state *SeatOpState) {
	x := s.StartX
	if s.Edges&proto.RiverWindowV1EdgesLeft != 0 {
		x += s.StartWidth - state.Window.Width
	}

	y := s.StartY
	if s.Edges&proto.RiverWindowV1EdgesTop != 0 {
		y += s.StartHeight - state.Window.Height
	}

	state.Window.SetPosition(x, y)
}

func (s *SeatOpResize) InformStart(w *Window) {
	w.ID.InformResizeStart()
}

func (s *SeatOpResize) InformEnd(w *Window) {
	w.ID.InformResizeEnd()
}

func NewSeatOpResize(w *Window, edges uint32) *SeatOpResize {
	return &SeatOpResize{
		StartX:      w.X,
		StartY:      w.Y,
		StartWidth:  w.Width,
		StartHeight: w.Height,
		Edges:       edges,
	}
}

type XkbBinding struct {
	proto.RiverXkbBindingV1Stub

	ID proto.RiverXkbBindingV1

	Seat     *Seat
	ActionFn func()
}

func (p *XkbBinding) HandleRiverXkbBindingV1Pressed(ctx context.Context) {
	p.Seat.PendingAction = p.ActionFn
}

func NewXkbBinding(seat *Seat, keysym uint32, mods uint32, fn func()) *XkbBinding {
	b := &XkbBinding{
		ID:       seat.WM.XkbBindingsV1.GetXkbBinding(seat.ID, keysym, mods),
		Seat:     seat,
		ActionFn: fn,
	}

	b.ID.SetUserData(b)
	return b
}

type PointerBinding struct {
	proto.RiverPointerBindingV1Stub

	ID proto.RiverPointerBindingV1

	Seat     *Seat
	ActionFn func()
}

func (p *PointerBinding) HandleRiverPointerBindingV1Pressed(ctx context.Context) {
	p.Seat.PendingAction = p.ActionFn
}

func NewPointerBinding(seat *Seat, button uint32, mods uint32, fn func()) *PointerBinding {
	b := &PointerBinding{
		ID:       seat.ID.GetPointerBinding(button, mods),
		Seat:     seat,
		ActionFn: fn,
	}

	b.ID.SetUserData(b)
	return b
}

type WindowManager struct {
	proto.WlRegistryStub
	proto.RiverWindowManagerV1Stub

	Registry        proto.WlRegistry
	WindowManagerV1 proto.RiverWindowManagerV1
	XkbBindingsV1   proto.RiverXkbBindingsV1

	Done bool
	Err  error

	Outputs []*Output
	Windows []*Window
	Seats   []*Seat
}

func (wm *WindowManager) HandleWlRegistryGlobal(ctx context.Context, name uint32, iface string, version uint32) {
	switch iface {
	case proto.RiverWindowManagerV1Name:
		if version >= 4 {
			wm.WindowManagerV1 = proto.As[proto.RiverWindowManagerV1](wm.Registry.Bind(name, iface, 4))
		}
	case proto.RiverXkbBindingsV1Name:
		wm.XkbBindingsV1 = proto.As[proto.RiverXkbBindingsV1](wm.Registry.Bind(name, iface, 1))
	}
}

func (wm *WindowManager) HandleRiverWindowManagerV1Unavailable(ctx context.Context) {
	wm.Done = true
	wm.Err = errors.New("another window manager is already running")
}

func (wm *WindowManager) HandleRiverWindowManagerV1Finished(ctx context.Context) {
	wm.Done = true
}

func (wm *WindowManager) HandleRiverWindowManagerV1Output(ctx context.Context, id proto.RiverOutputV1) {
	wm.Outputs = append(wm.Outputs, NewOutput(id))
}

func (wm *WindowManager) HandleRiverWindowManagerV1Window(ctx context.Context, id proto.RiverWindowV1) {
	wm.Windows = append(wm.Windows, NewWindow(id))
}

func (wm *WindowManager) HandleRiverWindowManagerV1Seat(ctx context.Context, id proto.RiverSeatV1) {
	wm.Seats = append(wm.Seats, NewSeat(id, wm))
}

func (wm *WindowManager) HandleRiverWindowManagerV1ManageStart(ctx context.Context) {
	wm.Outputs = slices.DeleteFunc(wm.Outputs, (*Output).MaybeDestroy)
	wm.Windows = slices.DeleteFunc(wm.Windows, (*Window).MaybeDestroy)
	wm.Seats = slices.DeleteFunc(wm.Seats, (*Seat).MaybeDestroy)

	for _, w := range wm.Windows {
		w.Manage()
	}
	for _, s := range wm.Seats {
		s.Manage()
	}

	wm.WindowManagerV1.ManageFinish()
}

func (wm *WindowManager) HandleRiverWindowManagerV1RenderStart(ctx context.Context) {
	for _, s := range wm.Seats {
		s.Render()
	}

	wm.WindowManagerV1.RenderFinish()
}

func run(ctx context.Context) (err error) {
	conn, err := wlcl.Connect(ctx, "")
	if err != nil {
		return fmt.Errorf("connect: %w", err)
	}
	defer func() {
		if cerr := conn.Close(); cerr != nil && err == nil {
			err = cerr
		}
	}()

	// Don't pass WAYLAND_DEBUG on to children, the added noise makes
	// debugging the window manager itself impractical.
	// It only matters if it's set when the display is created.
	os.Unsetenv("WAYLAND_DEBUG")

	display := proto.As[proto.WlDisplay](conn.Init(proto.Interfaces))

	wm := &WindowManager{}

	wm.Registry = display.GetRegistry()
	wm.Registry.SetUserData(wm)

	if err := wlcl.Roundtrip(ctx, display); err != nil {
		return fmt.Errorf("roundtrip: %w", err)
	}

	if !wm.WindowManagerV1.IsSet() || !wm.XkbBindingsV1.IsSet() {
		return fmt.Errorf("no required globals")
	}

	wm.WindowManagerV1.SetUserData(wm)

	for !wm.Done {
		if err := conn.Dispatch(ctx); err != nil {
			return fmt.Errorf("dispatch: %w", err)
		}
	}

	return wm.Err
}

func main() {
	ctx := context.Background()
	if err := run(ctx); err != nil {
		log.Fatal(err)
	}
}

File: lua/tinyrwm.lua
#! /usr/bin/env lua
-- SPDX-FileCopyrightText: © 2026 FireFly
-- SPDX-License-Identifier: 0BSD

local wau = require("wau")
local xkbcommon = require("tinyrwm.xkbcommon")
local posix = require("posix")

wau:require("tinyrwm.protocol.river-window-management-v1")
wau:require("tinyrwm.protocol.river-xkb-bindings-v1")

local globals = {}
local required_globals = {
    ["river_window_manager_v1"] = 4,
    ["river_xkb_bindings_v1"] = 1,
}

local Mods = wau.river_seat_v1.Modifiers

local xkb_bindings = {
    {"space", Mods.MOD4, "spawn-foot"},
    {"q", Mods.MOD4, "close"},
    {"n", Mods.MOD4, "focus-next"},
    {"Escape", Mods.MOD4, "exit"},
}

local pointer_bindings = {
    {"left", Mods.MOD4, "move"},
    {"right", Mods.MOD4, "resize"},
}

local wm = {
    outputs = {},
    seats = {},
    -- Windows are kept in rendering order; last window is topmost
    windows = {},
}

local function table_index_of(tbl, sought)
    for i, v in ipairs(tbl) do
        if v == sought then return i end
    end
    return 0
end

local function table_filter_inplace(tbl, pred)
    local removed = 0
    for i=1,#tbl do
        if pred(tbl[i]) then
            tbl[i - removed] = tbl[i]
        else
            removed = removed + 1
        end
        if removed > 0 then tbl[i] = nil end
    end
    return tbl
end


---- Output ---------------------------
local Output = { mt = {}, listener = {} }
Output.mt.__index = Output

function Output.create(obj)
    local output = { obj = obj }
    setmetatable(output, Output.mt)
    obj:set_user_data(output)
    obj:add_listener(Output.listener)
    return output
end

function Output:maybe_destroy()
    if self.removed then
        self.obj:destroy()
    else
        return self
    end
end

function Output.listener:removed()
    self:get_user_data().removed = true
end


---- Window ---------------------------
local Window = { mt = {}, listener = {} }
Window.mt.__index = Window

function Window.create(obj)
    local window = {
        obj = obj,
        node = obj:get_node(),
        new = true,
    }
    setmetatable(window, Window.mt)
    obj:set_user_data(window)
    obj:add_listener(Window.listener)
    return window
end

function Window:maybe_destroy()
    if self.closed then
        self.obj:destroy()
        self.node:destroy()
    else
        return self
    end
end

function Window:manage()
    if self.new then
        self.new = nil
        self:set_position(0, 0)
        self.obj:propose_dimensions(0, 0)
    end

    local move = self.pointer_move_requested
    if move ~= nil then
        self.pointer_move_requested = nil
        move.seat:pointer_move(self)
    end

    local resize = self.pointer_resize_requested
    if resize ~= nil then
        self.pointer_resize_requested = nil
        resize.seat:pointer_resize(self, resize.edges)
    end
end

function Window:set_position(x, y)
    self.node:set_position(x, y)
    self.x = x
    self.y = y
end

function Window.listener:closed()
    self:get_user_data().closed = true
end
function Window.listener:dimensions(width, height)
    local window = self:get_user_data()
    window.width = width
    window.height = height
end
function Window.listener:pointer_move_requested(seat)
    self:get_user_data().pointer_move_requested = {
        seat = seat:get_user_data(),
    }
end
function Window.listener:pointer_resize_requested(seat, edges)
    local Edges = wau.river_window_v1.Edges
    self:get_user_data().pointer_resize_requested = {
        seat = seat:get_user_data(),
        edges = {
            left = (edges & Edges.LEFT) ~= 0,
            right = (edges & Edges.RIGHT) ~= 0,
            top = (edges & Edges.TOP) ~= 0,
            bottom = (edges & Edges.BOTTOM) ~= 0,
        },
    }
end


---- Seat -----------------------------
local Seat = { mt = {}, listener = {} }
Seat.mt.__index = Seat

function Seat.create(obj)
    local seat = {
        obj = obj,
        new = true,
        xkb_bindings = {},
        pointer_bindings = {},
    }
    setmetatable(seat, Seat.mt)
    obj:set_user_data(seat)
    obj:add_listener(Seat.listener)
    return seat
end

function Seat:focus(window)
    if window == nil and #wm.windows > 0 then
        -- Fall back to topmost window
        window = wm.windows[#wm.windows]
    end

    if window then
        if self.focused ~= window then
            self.obj:focus_window(window.obj)
            self.focused = window
            -- Move to top
            local i = table_index_of(wm.windows, window)
            table.remove(wm.windows, i)
            table.insert(wm.windows, window)
            window.node:place_top()
        end
    else
        self.obj:clear_focus()
        self.focused = nil
    end
end

function Seat:pointer_move(window)
    if self.op == nil then
        self:focus(window)
        self.obj:op_start_pointer()
        self.op = {
            type = "move",
            window = window,
            start = { x = window.x, y = window.y },
            dx = 0,
            dy = 0,
        }
    end
end

function Seat:pointer_resize(window, edges)
    if self.op == nil then
        self:focus(window)
        window.obj:inform_resize_start()
        self.obj:op_start_pointer()
        self.op = {
            type = "resize",
            window = window,
            edges = edges,
            start = {
                x = window.x,
                y = window.y,
                width = window.width,
                height = window.height,
            },
            dx = 0,
            dy = 0,
        }
    end
end

function Seat:action(action)
    if action == "spawn-foot" then
        if posix.unistd.fork() == 0 then
            posix.unistd.execp("foot", {})
        end
    elseif action == "close" then
        if self.focused ~= nil then
            self.focused.obj:close()
        end
    elseif action == "focus-next" then
        self:focus(wm.windows[1])
    elseif action == "move" then
        if self.hovered ~= nil then
            self:pointer_move(self.hovered)
        end
    elseif action == "resize" then
        if self.hovered ~= nil then
            self:pointer_resize(self.hovered, { bottom = true, right = true })
        end
    elseif action == "exit" then
        globals["river_window_manager_v1"]:exit_session()
    else
        print("Seat:action: unimplemented", action)
    end
end

function Seat:add_pointer_binding(button, mods, action)
    -- From /usr/include/linux/input-event-codes.h
    local button_code = ({ left = 0x110, right = 0x111 })[button]
    local obj = self.obj:get_pointer_binding(button_code, mods)
    local binding = { obj = obj }

    obj:add_listener {
        ["pressed"] = function (_)
            self.pending_action = action
        end,
    }
    obj:enable()
    table.insert(self.pointer_bindings, binding)
end

function Seat:add_xkb_binding(key, mods, action)
    local keysym = xkbcommon.keysym(key)
    local obj = globals["river_xkb_bindings_v1"]:get_xkb_binding(
                    self.obj, keysym, mods)
    local binding = { obj = obj }

    obj:add_listener {
        ["pressed"] = function (_)
            self.pending_action = action
        end,
    }
    obj:enable()
    table.insert(self.xkb_bindings, binding)
end

function Seat:manage()
    if self.new then
        self.new = nil

        for _, tbl in ipairs(xkb_bindings) do
            self:add_xkb_binding(table.unpack(tbl))
        end

        for _, tbl in ipairs(pointer_bindings) do
            self:add_pointer_binding(table.unpack(tbl))
        end
    end

    if self.focused and self.focused.closed then
        self.focused = nil
    end

    self:focus(self.interacted)
    self.interacted = nil

    if self.pending_action ~= nil then
        self:action(self.pending_action)
        self.pending_action = nil
    end

    if self.op and self.op.window then
        local op, window = self.op, self.op.window
        local window = self.op.window

        if window.closed then
            self.obj:op_end()
            self.op = nil

        elseif self.op_release then
            if op.type == "resize" then
                window.obj:inform_resize_end()
            end
            self.obj:op_end()
            self.op = nil

        elseif op.type == "resize" then
            local width = math.max(
                1,
                op.edges.left and (op.start.width - op.dx) or
                op.edges.right and (op.start.width + op.dx) or
                op.start.width
            )
            local height = math.max(
                1,
                op.edges.top and (op.start.height - op.dy) or
                op.edges.bottom and (op.start.height + op.dy) or
                op.start.height
            )
            window.obj:propose_dimensions(width, height)
        end
    end

    self.op_release = nil
end

function Seat:render()
    if self.op and self.op.window then
        local op, window = self.op, self.op.window
        if self.op.type == "move" then
            window:set_position(
                op.start.x + op.dx,
                op.start.y + op.dy
            )
        elseif self.op.type == "resize" then
            local x = op.edges.left
                      and (op.start.x + (op.start.width - window.width))
                      or op.start.x
            local y = op.edges.top
                      and (op.start.y + (op.start.height - window.height))
                      or op.start.y
            window:set_position(x, y)
        end
    end
end

function Seat:maybe_destroy()
    if self.removed then
        for _, binding in ipairs(self.xkb_bindings) do
            binding.obj:destroy()
        end
        for _, binding in ipairs(self.pointer_bindings) do
            binding.obj:destroy()
        end
        self.obj:destroy()
    else
        return self
    end
end

function Seat.listener.removed(self)
    self:get_user_data().removed = true
end
function Seat.listener.pointer_enter(self, window)
    self:get_user_data().hovered = window:get_user_data()
end
function Seat.listener.pointer_leave(self)
    self:get_user_data().hovered = nil
end
function Seat.listener.window_interaction(self, window)
    self:get_user_data().interacted = window:get_user_data()
end
function Seat.listener.op_delta(self, dx, dy)
    local seat = self:get_user_data()
    seat.op.dx = dx
    seat.op.dy = dy
end
function Seat.listener.op_release(self)
    self:get_user_data().op_release = true
end


---- wm -------------------------------
local function wm_manage()
    table_filter_inplace(wm.outputs, Output.maybe_destroy)
    table_filter_inplace(wm.windows, Window.maybe_destroy)
    table_filter_inplace(wm.seats, Seat.maybe_destroy)

    for _, window in ipairs(wm.windows) do
        window:manage()
    end

    for _, seat in ipairs(wm.seats) do
        seat:manage()
    end

    globals["river_window_manager_v1"]:manage_finish()
end

local function wm_render()
    for _, seat in ipairs(wm.seats) do
        seat:render()
    end

    globals["river_window_manager_v1"]:render_finish()
end

local wm_handlers = {
    ["unavailable"] = function (self)
        io.stderr:write("another window manager is already running\n")
        os.exit(1)
    end,
    ["finished"] = function (self)
        os.exit(0)
    end,
    ["manage_start"] = wm_manage,
    ["render_start"] = wm_render,
    ["output"] = function (self, obj)
        table.insert(wm.outputs, Output.create(obj))
    end,
    ["seat"] = function (self, obj)
        table.insert(wm.seats, Seat.create(obj))
    end,
    ["window"] = function (self, obj)
        table.insert(wm.windows, Window.create(obj))
    end,
}


---- Entry point ----------------------
display = wau.wl_display.connect()
assert(display, "Failed to connect to wayland compositor")

-- Ensure we exit nonzero if an event handler errors
local function handle_callback_error(proxy, name, func, err)
    io.stderr:write(("-- Error calling event handler for %s %q:")
                        :format(tostring(proxy), name))
    io.stderr:write(("%s\n"):format(tostring(err)))
    os.exit(1)
end
wau.wl_proxy.set_error_callback(handle_callback_error)

-- Avoid passing WAYLAND_DEBUG to our children
posix.stdlib.setenv("WAYLAND_DEBUG", nil)

-- Ensure children are automatically reaped
posix.signal.signal(posix.signal.SIGCHLD, posix.signal.SIG_IGN)

local registry = display:get_registry()
registry:add_listener {
    ["global"] = function (self, name, iface, version)
        local required_version = required_globals[iface]
        if required_version ~= nil then
            assert(required_version <= version,
                ("wayland compositor supported %s version too old (need %d, got %d)")
                    :format(iface, required_version, version))
            globals[iface] = self:bind(name, wau[iface], required_version)
        end
    end,
}

display:roundtrip()

for k in pairs(required_globals) do
    assert(globals[k] ~= nil, ("wayland compositor does not support %s"):format(k))
end

globals["river_window_manager_v1"]:add_listener(wm_handlers)

while display:dispatch() do end
